core/arch/arch.c

```diff
@@ -3355,45 +3355,59 @@ dr_mcontext_to_priv_mcontext(priv_mcontext_t *dst, dr_mcontext_t *src)
      * if we append to dr_mcontext_t in the future we'll need
      * to check src->size here.
      */
-    if (src->size != sizeof(dr_mcontext_t) &&
-        src->size != sizeof(dr_mcontext_t) - sizeof(dr_opmask_t))
+    if (src->size > sizeof(dr_mcontext_t))
         return false;
-    if (TESTALL(DR_MC_ALL, src->flags))
+    if (TEST(DR_MC_ALL, src->flags) &&
+        IF_X86_ELSE(src->size > offsetof(priv_mcontext_t, opmask), true)) {
         *dst = *(priv_mcontext_t *)(&MCXT_FIRST_REG_FIELD(src));
-    else {
-        if (TEST(DR_MC_INTEGER, src->flags)) {
+    } else {
+        if (TESTANY(DR_MC_INTEGER | DR_MC_ALL, src->flags)) {
             /* xsp is in the middle of the mcxt, so we save dst->xsp here and
              * restore it later so we can use one memcpy for DR_MC_INTEGER.
              */
             reg_t save_xsp = dst->xsp;
-            memcpy(&MCXT_FIRST_REG_FIELD(dst), &MCXT_FIRST_REG_FIELD(src),
-                   /* end of the mcxt integer gpr */
-                   offsetof(priv_mcontext_t, IF_X86_ELSE(xflags, pc)));
+            if (src->size >= offsetof(priv_mcontext_t, IF_X86_ELSE(xflags, pc))) {
+                memcpy(&MCXT_FIRST_REG_FIELD(dst), &MCXT_FIRST_REG_FIELD(src),
+                       /* end of the mcxt integer gpr */
+                       offsetof(priv_mcontext_t, IF_X86_ELSE(xflags, pc)));
+            }
             dst->xsp = save_xsp;
         }
-        if (TEST(DR_MC_CONTROL, src->flags)) {
+        if (TESTANY(DR_MC_CONTROL | DR_MC_ALL, src->flags)) {
             /* XXX i#2710: mc->lr should be under DR_MC_CONTROL */
             dst->xsp = src->xsp;
-            dst->xflags = src->xflags;
-            dst->pc = src->pc;
+            if (src->size > offsetof(priv_mcontext_t, xflags))
+                dst->xflags = src->xflags;
+            if (src->size > offsetof(priv_mcontext_t, pc))
+                dst->pc = src->pc;
         }
-        if (TEST(DR_MC_MULTIMEDIA, src->flags)) {
+        if (TESTANY(DR_MC_MULTIMEDIA | DR_MC_ALL, src->flags)) {
             /* XXX i#1312: The structure size will double on 64-bit UNIX (possibly
-             * Windows) builds. A corresponding size check be will added with a
+             * Windows) builds. A corresponding size check will be added with a
              * future patch.
              */
-            IF_X86_ELSE({ memcpy(&dst->simd, &src->simd, sizeof(dst->simd)); },
-                        {
-                            /* FIXME i#1551: NYI on ARM */
-                            ASSERT_NOT_IMPLEMENTED(false);
-                        });
-            if (src->size > sizeof(dr_mcontext_t) - sizeof(dr_opmask_t)) {
-                IF_X86_ELSE({ memcpy(&dst->opmask, &src->opmask, sizeof(dst->opmask)); },
-                            {
-                                /* FIXME i#1551: NYI on ARM */
-                                ASSERT_NOT_IMPLEMENTED(false);
-                            });
-            }
+            IF_X86_ELSE(
```

Comment:
IMHO a regular ifdef is more readable at this point

Reply:
Done.